modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +162,141 @@ impl VulnerabilityService {
+
+                let version = match purl.version.as_ref() {
+                    Some(v) => v,
+                    None => return Err(Error::Purl(PurlErr::MissingVersion(p.to_string()))),
+                };
+
+                let ns_condition = match purl.namespace.clone() {
+                    Some(namespace) => {
+                        let sql = "base_purl.namespace = $1";
+                        Statement::from_sql_and_values(
+                            connection.get_database_backend(),
+                            sql,
+                            [namespace.into()],
+                        )
+                    }
+                    None => Statement::from_string(
+                        connection.get_database_backend(),
+                        "base_purl.namespace IS NULL",
+                    ),
+                };
+
+                let name = purl.name;
+                let ty = purl.ty;
+                let values = [
+                    p.to_string().into(),
+                    name.into(),
+                    ty.into(),
+                    version.to_string().into(),
+                ];
+                let sql = format!(r#"
+                  SELECT $1 as requested_purl,
+                    vulnerability.id, vulnerability.title, vulnerability.reserved,
+                    vulnerability.published, vulnerability.modified, vulnerability.withdrawn, vulnerability.cwes
+                  FROM base_purl
+                    LEFT JOIN purl_status ON base_purl.id = purl_status.base_purl_id
+                    INNER JOIN version_range ON purl_status.version_range_id = version_range.id
+                    LEFT JOIN vulnerability ON purl_status.vulnerability_id = vulnerability.id
+                    INNER JOIN status ON purl_status.status_id = status.id
+                  WHERE {ns_condition}
+                    AND base_purl.name = $2
+                    AND base_purl.type = $3
+                    AND version_matches($4, version_range.*) = TRUE
+                    AND status.slug != 'fixed'"#);
+                let query =
+                    Statement::from_sql_and_values(connection.get_database_backend(), &sql, values);
+                Ok(query.to_string())
+            })
+            .collect();
+
+        let mut errors = Vec::new();
+
+        let query = queries
+            .into_iter()
+            .filter_map(|x| match x {
+                Ok(r) => Some(r),
+                Err(e) => {
+                    errors.push(e);
+                    None
+                }
+            })
+            .collect::<Vec<String>>()
+            .join(" UNION ALL ");
+
+        if !errors.is_empty() {
+            return Err(Error::BadRequest(format!(
+                "Invalid purls received: [{}]",
+                errors
+                    .iter()
+                    .map(|e| e.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )));
+        }
```

Comment:
I don't love constructing errors like this in the code. I'd rather use macros in the Error struct, but in this case, I think the `BadRequest` is a little overkill. I think the common case is that all the purl's will be valid, so it's fine to return one invalid one at a time. If you think it's more common than I think, and the caller would prefer to handle all at once rather than one at a time, you can leave it, but construct the mutable errors array at the top of the fn and use map_err to fill it like the pattern I linked to above, please.
```suggestion
```